### ISSUE TITLE
Implement manual garbage collection in content build

### ISSUE THREAD

**Issue by alexandec** [department-of-veterans-affairs/va.gov-team#25074]
Discussed in [#24022](https://app.zenhub.com/workspaces/vsp---frontend-tools-5fc9325744944e0015ed1861/issues/department-of-veterans-affairs/va.gov-team/24022).

### Details

Parts of the content build rapidly allocate and deallocate large amounts of data. V8's garbage collection handles this well in most cases, but it can fall behind. This causes spikes in heap and non-heap (buffer) allocation, resulting in very large peak memory use at high node counts (see the above ticket for more information).

Triggering manual garbage collection at a 10-second interval will keep memory use in check, particularly during the Apply Layouts steps. In addition, writing total memory use to the log will be a useful way to track changes due to garbage collection, and to track memory use over time.

### Acceptance Criteria

- [x] Manual garbage collection is performed at a regular interval
- [x] Total memory use (RSS) is logged to allow monitoring over time

**Comment by alexandec** [comment]
#### Garbage collection interval findings

I tested the content build on Jenkins with multiple garbage collection intervals to find the best optimization between build time and memory use. For context, a 10-second GC interval means we're manually triggering a mark-and-sweep GC (the more time-consuming type) every 10 seconds. 

| GC Interval | Build Time | Peak Memory |
| ----------- | ----------- | --------- |
| 10 seconds | 52 minutes | 6.7 gb     |
| 20 seconds | 49 minutes | 6.9 gb    |
| 30 seconds | 48 minutes  | 7.0 gb   |
| 40 seconds | 58 minutes   | 9.0gb     |
| 50 seconds | 52 minutes | 9.4gb | 
| 60 seconds | 55 minutes | 9.9 gb    |

However, I found that while a 30-second interval is optimal in theory, in practice it allowed occasional large increases in memory use during the "Apply Layouts" and "Parse a virtual DOM" build steps. These memory spikes didn't happen on every build, but they were frequent enough to make build failure a risk. The 10-second interval is a safer choice even though it adds 3 or 4 minutes to the build time.

**Comment by alexandec** [comment]
Here's the Prometheus [free memory graph](http://prometheus-utility.vfs.va.gov:9090/prometheus/graph?g0.range_input=1h&g0.expr=node_memory_MemAvailable%7Bjob%3D%22dsva-vetsgov-utility-jenkins-worker-node%22%7D&g0.tab=0) from a Jenkins run (dev, staging, and prod content builds in parallel) with 48k nodes, 10 second GC interval, 5gb heap limit (max-old-space-size):

![prometheus-48k-10seconds-5gb.png](https://images.zenhubusercontent.com/5fc6638ec869251f539d3881/9d078f15-4c27-48ef-825d-10719c5ec02e)

Manual garbage collection at a regular interval keeps memory usage spikes minimal throughout the build. Without it, the spikes shown would be much larger, in some cases using up all available memory and causing Jenkins to kill the process. Or the heap would fill and V8 would throw an error. Note that this run (and the above runs) are with debug information turned off. Enabling debug information would also use up all available memory. We have [another ticket ](https://app.zenhub.com/workspaces/vsp---frontend-tools-5fc9325744944e0015ed1861/issues/department-of-veterans-affairs/va.gov-team/25100) to add debug information at the end of the build in a safer manner.

Here's a summary of memory usage and elapsed time:

![Screen Shot 2021-06-04 at 1.56.40 PM.png](https://images.zenhubusercontent.com/5fc6638ec869251f539d3881/4e43e83c-af94-41e6-b5e3-02cfb2b3c8d3)

I added RSS (total memory use) to these stats so we can track it more easily. I also added tracking for peak RSS at 10 second intervals to get a more accurate value, instead of just reporting it at the end of each build step. The final peak value is logged at the end of the entire build.
